src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -621,19 +616,14 @@ internal virtual T ReadAsPropertyNameCore(ref Utf8JsonReader reader, Type typeTo
         /// <remarks>Method should be overridden in custom converters of types used in serialized dictionary keys.</remarks>
         public virtual void WriteAsPropertyName(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
         {
-            if (!IsInternalConverter &&
-                options.SerializerContext is null && // For consistency do not return any default converters for
-                                                     // options instances linked to a JsonSerializerContext,
-                                                     // even if the default converters might have been rooted.
-                DefaultJsonTypeInfoResolver.TryGetDefaultSimpleConverter(TypeToConvert, out JsonConverter? defaultConverter))
+            // .NET 5 backward compatibility: hardcode the default converter for primitive key serialization.
+            JsonConverter<T>? fallbackConverter = GetFallbackConverterForPropertyNameSerialization(options);
+            if (fallbackConverter is null)
             {
-                // .NET 5 backward compatibility: hardcode the default converter for primitive key serialization.
-                Debug.Assert(defaultConverter.IsInternalConverter && defaultConverter is JsonConverter<T>);
-                ((JsonConverter<T>)defaultConverter).WriteAsPropertyNameCore(writer, value, options, isWritingExtensionDataProperty: false);
-                return;
+                ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
             }
 
-            ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
+            fallbackConverter.WriteAsPropertyNameCore(writer, value, options, isWritingExtensionDataProperty: false);
```

Comment:
Is this branch actually ever hit?

Reply:
Why not? the resultant fallback converter may or may not be null.

Reply:
Nvm I had a confusion from https://github.com/dotnet/runtime/pull/77488/files#r1007381913.